docs/design/mono/embedding-api-deprecation.md

```diff
@@ -0,0 +1,68 @@
+
+## ABI stability
+
+ABI stability is only guaranteed across major version releases. .NET 6 will not receive a servicing update that breaks ABI compatibility, but you cannot link a .NET 5 app against the .NET 6 release, full stop.
+
+## API stability
+
+API stability is more complicated. There are two distinct cases here: .NET 6 and the behavior moving forward.
+
+With the .NET 6 release specifically, we have an opportunity to do some initial API cleanup work because relatively few desktop users have migrated over to .NET 5 Mono. There is some low-hanging fruit to potentially clean up immediately, before we have significant migration: irrelevant functionality like `mono_domain_set` or `refonly` parameters, legacy compatibility hacks like `MonoLookupPInvokeStatus`, and subtly-incorrect behavior like fixed-size GC handles. 
+
+The .NET 6 immediate removal should only be used for functions that will allow for significant cleanup of the embedding API, runtime internals, or that would otherwise be likely to result in difficult to diagnose customer bugs or crashes at runtime. There are a lot of functions we might not want to remove for .NET 6 for ease of migration or simply time constraints. The earlier `mono_assembly_load` example is a potential case of this, but we have many similar cases of symbol variants. We want to remove many of them eventually, but for one of the above reasons it may not happen for .NET 6.
+
+To handle this moving forward, we will use a `MONO_DEPRECATED` define, which corresponds to the `[[deprecated]]` attribute on relevant compilers. For every API marked with this, we should have an entry on a new page on the Mono site explaining the deprecation and the suggested alternative. This creates a moderate bar for deprecations, but still leaves them as a possibility where necessary.
+
+Any function appearing in a major release marked deprecated may, and likely will go away in the following major release. This means that if `mono_assembly_load` is deprecated for .NET 6, it can be removed at any time leading up to .NET 7 or beyond, including during the previews.
+
+Deprecation is the preferred path for functions likely to see heavy use with embedders in .NET 6 and the _only_ option past that point. .NET 6 is the only release where removal may occur without the attribute first being present in a major release, though this should be avoided with functions likely to see heavy use in existing embedding workloads.
+
+# Points for further discussion (to address before merging)
+
+For deprecated APIs that include useless arguments or are irrelevant for netcore, should we silently ignore them/do nothing or throw a runtime error? What would be more useful for embedders, given that the deprecations allow them to detect potentially problematic functions at compile time?
+
+How does this work with LTS releases? Should we try and deprecate on LTS and remove on non-LTS when possible, or does it not matter?
```

Comment:
removing symbols in LTS releases feels wrong to me.

Reply:
I don't personally think it's a problem. If you're only using LTS releases, you'll get hit with the removal regardless of whether it was done on the non-LTS release or not. However, I put this down to get more input since I know you feel differently :)

Reply:
You're right, if I'm working on a project that's tracking LTS releases and I jump from one LTS to the next then the porting burden is the same no matter when we deprecate a function.

I guess the scenario I'm worried about is that it doesn't leave us a good way to react if we remove a symbol and find out that we made a mistake and didn't cover every use by an adequate replacement.  With an non-LTS release it's only ~12months of pain (assuming we can't fix it in a servicing release).  With LTS it's 24 months if they can jump to the subsequent LTS release early, or 36 if they have to wait for their adopted release to EOL.

Reply:
The only real difference between LTS and non-LTS is in the length of support, ie in what happens after the release ships.

We had similar discussions about whether it is a good idea to push more breaking changes in non-LTS releases or not. The conclusion of these discussions has always been that we have to keep the bar for quality or breaking changes constant, independent of whether the current release is LTS or non-LTS.